### ISSUE TITLE
[JavaScript] Implement new Concept Exercise: closures

### ISSUE THREAD

**Issue by SleeplessByte** [exercism/v3#1890]
This issue describes how to implement the `closures` concept exercise for the `javascript` track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

See the documentation _above_ (general documentation), as well as [How to implement a Concept Exercise in JavaScript](https://github.com/exercism/v3/blob/master/languages/javascript/reference/implementing-a-concept-exercise.md).

## Goal

The goal of this exercise is to teach the student how _closures_ are implemented in `JavaScript`. 

In other words: how _values_ can be _enclosed_ in a _function_. It touches onto `scoping` and `functions` in general, and is usually used as `higher-order-functions`, `callbacks` or to change `visibility` (think: private).

## JavaScript specificness

ℹ You might be inclined to think that this is equivalent to **anonymous functions**, but that is not the case. In most languages I know supporting closures, these are modelled as anonymous functions (functions without a name), such as inline blocks, proc, lambda or even function definitions. The JavaScript equivalent is:

```javascript
// As function argument
someFunctionCall(() => { /* my anonymous arrow function declaration */ })

// As return value
function anotherFunctionCall() {
  return function () {
    /* anonymous function declaration */
  }
)
```

However, in JavaScript, the function name can be derived from the _original binding_. This is **not common*. What does this mean? The following examples are **not** anonymous functions:

``` javascript
const adder = (a, b) => a + b

adder.name
// => "adder"


let value = 0
var increment = function () {
  value += 1
  return value
}

increment.name
// "increment"

const rebound  = increment
rebound.name
// "increment"
```

So in JavaScript, the anonymous-ness doesn't determine if you can _enclose_ something. In fact, you can _always_ enclose values, regardless if you're using a:

- function declaration
- anonymous function declaration
- function expression
- anonymous function expression
- `Function` object/instance/class with evaluation*
- `const`, `let` or `var`

```javascript
let a = 'hello world'
const print = new Function('console.log(a)')

print.name
// => "anonymous"

print()
// => prints "hello world"
```

## Learning objectives

- Function that returns an "enclosed" value (think `private`)
- Function that _mutates_ an "enclosed" value (think `increment`/`id` generation)
- Function that combines an "enclosed" value with some input (think higher order/adder)
- Variable scope
- Creating closures in loops (A common mistake)

## Out of scope

- Shadowing values (unless you can make this very concise)
- Hoisting (especially: enclosing `var`)
- Context (`this`, `arguments`, `new.target`)

## Concepts

- `closures`

## Prerequisites

- `basics`
- ... (do you think something is missing?)

## Resources to refer to

- [MDN: Closures](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures)
- ...

### After

You can go wild here. It's a very interesting subject. I would definitely try to do use-cases.

## Implementing

I recommend forking [`elixir/anonymous-functions`](https://github.com/exercism/v3/blob/master/languages/elixir/exercises/concept/anonymous-functions/.docs/instructions.md
). 

## Help

You can choose to do this solo-style, or collaborate with multiple people on this. My suggestion would be is to 

1. **first accept this issue by saying "I'd like to work on this"** (you will not get a response from me. Just go with it) and optionally request that someone works with you (and wait for a second person to accept your request).
2. Use this issue to write down what you need (collect reference material) and discuss, if necessary, what to write and what not to write.
3. Draft a PR and request the feedback from at least one other JavaScript member (I recommend just pinging the channel on slack).
